### PR TITLE
[Feature]Add patch embed and patch merge

### PR DESCRIPTION
**Pull request by jshilong**
## Motivation

`PatchEmbed` and `PatchMerging` are basic modules that have been used in many transformer methods, such as VIT, PVT, and Swin
At the same time, `AdaptivePadding` is an essential feature to deal with the dynamic shape in Detection and Segmentation, which avoids losing pixels of the feature map.

## Modification

- Add `AdaptivePadding`, `PatchEmbed` and `PatchMerging`  to `/mmcv/cnn/bricks/transformer.py` and add the corresponding unitests.
- Always apply adaptive padding first before doing the `PatchEmbed` and `PatchMerging` to deal with the dynamic shape in Detection and Segmentation.

## BC-breaking (Optional)
None

### CODE REVIEW

File: mmcv/cnn/bricks/transformer.py [hunk 1]
```diff
@@ -55,6 +59,332 @@ def build_transformer_layer_sequence(cfg, default_args=None):
     return build_from_cfg(cfg, TRANSFORMER_LAYER_SEQUENCE, default_args)
 
 
+class AdaptivePadding(nn.Module):
+    """Applies padding to input (if needed) so that input can get fully covered
```
Comment (ZwwWayne) [inline mmcv/cnn/bricks/transformer.py:63]: The summary of this module is too long. Only keep one sentence as summary and leave others as description.

File: mmcv/cnn/bricks/transformer.py [hunk 2]
```diff
@@ -55,6 +59,332 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+        >>>     padding="corner")
+        >>> out = adap_pad(input)
+        >>> assert (out.shape[2], out.shape[3]) == (16, 32)
+        >>> input = torch.rand(1, 1, 16, 17)
+        >>> out = adap_pad(input)
+        >>> assert (out.shape[2], out.shape[3]) == (16, 32)
+    """
+
+    def __init__(self, kernel_size=1, stride=1, dilation=1, padding='corner'):
+
```
Comment (ZwwWayne) [inline mmcv/cnn/bricks/transformer.py:94]: This empty line is unnecessary

File: mmcv/cnn/bricks/transformer.py [hunk 3]
```diff
@@ -55,6 +59,332 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+        output_h = math.ceil(input_h / stride_h)
+        output_w = math.ceil(input_w / stride_w)
+        pad_h = max((output_h - 1) * stride_h +
+                    (kernel_h - 1) * self.dilation[0] + 1 - input_h, 0)
+        pad_w = max((output_w - 1) * stride_w +
+                    (kernel_w - 1) * self.dilation[1] + 1 - input_w, 0)
+        return pad_h, pad_w
+
+    def forward(self, x):
+        pad_h, pad_w = self.get_pad_shape(x.size()[-2:])
```
Comment (ZwwWayne) [inline mmcv/cnn/bricks/transformer.py:122]: docstring.

File: mmcv/cnn/bricks/transformer.py [hunk 4]
```diff
@@ -55,6 +59,332 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+        padding = to_2tuple(padding)
+        dilation = to_2tuple(dilation)
+
+        self.padding = padding
+        self.kernel_size = kernel_size
+        self.stride = stride
+        self.dilation = dilation
+
+    def get_pad_shape(self, input_shape):
+        input_h, input_w = input_shape
```
Comment (ZwwWayne) [inline mmcv/cnn/bricks/transformer.py:110]: docstring.

File: mmcv/cnn/bricks/transformer.py [hunk 5]
```diff
@@ -55,6 +59,332 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+
+class PatchEmbed(BaseModule):
+    """Image to Patch Embedding.
+
+    We use a conv layer to implement PatchEmbed.
+
+    Args:
+        in_channels (int): The num of input channels. Default: 3
+        embed_dims (int): The dimensions of embedding. Default: 768
+        conv_type (str): The config dict for embedding
```
Comment (Junjun2016) [inline mmcv/cnn/bricks/transformer.py:142]: Not `config dict` (str).

File: mmcv/cnn/bricks/transformer.py [hunk 6]
```diff
@@ -55,6 +59,332 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+            conv layer type selection. Default: "Conv2d.
```
Comment (Junjun2016) [inline mmcv/cnn/bricks/transformer.py:143]:
```suggestion
            conv layer type selection. Default: "Conv2d".
```

File: mmcv/cnn/bricks/transformer.py [hunk 7]
```diff
@@ -55,6 +59,332 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+            Default: None.
+    """
+
+    def __init__(
+        self,
+        in_channels=3,
+        embed_dims=768,
+        conv_type='Conv2d',
+        kernel_size=16,
+        stride=16,
```
Comment (Junjun2016) [inline mmcv/cnn/bricks/transformer.py:168]: The default value is not consistent with docstring.

File: mmcv/cnn/bricks/transformer.py [hunk 8]
```diff
@@ -55,6 +59,332 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+
+    This layer groups feature map by kernel_size, and applies norm and linear
+    layers to the grouped feature map. Our implementation uses `nn.Unfold` to
+    merge patch, which is about 25% faster than original implementation.
+    Instead, we need to modify pretrained models for compatibility.
+
+    Args:
+        in_channels (int): The num of input channels.
+            to gets fully covered by filter and stride you specified..
+            Default: True.
```
Comment (Junjun2016) [inline mmcv/cnn/bricks/transformer.py:271]: Remove.

File: mmcv/cnn/bricks/transformer.py [hunk 9]
```diff
@@ -55,6 +59,332 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+        embed_dims=768,
+        conv_type='Conv2d',
+        kernel_size=16,
+        stride=16,
+        padding='corner',
+        dilation=1,
+        bias=True,
+        norm_cfg=None,
+        input_size=None,
+        init_cfg=None,
```
Comment (Junjun2016) [inline mmcv/cnn/bricks/transformer.py:174]:
```suggestion
        init_cfg=None
```

File: tests/test_cnn/test_transformer.py [hunk 10]
```diff
@@ -2,11 +2,469 @@
...
+        pool = AdaptivePadding(
+            kernel_size=kernel_size,
+            stride=stride,
+            dilation=dilation,
+            padding=padding)
+        out = pool(input)
+        # padding to divisible by 16
+        assert (out.shape[2], out.shape[3]) == (16, 32)
+        input = torch.rand(1, 1, 16, 17)
+        out = pool(input)
```
Comment (Junjun2016) [inline tests/test_cnn/test_transformer.py:28]: `pool` --> `adap_pad`

File: mmcv/cnn/bricks/transformer.py [hunk 11]
```diff
@@ -55,6 +59,332 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+    """Applies padding to input (if needed) so that input can get fully covered
+    by filter you specified. It support two modes "same" and "corner". The
+    "same" mode is same with "SAME" padding mode in TensorFlow, pad zero around
+    input. The "corner"  mode would pad zero to bottom right.
+
+    Args:
+        kernel_size (int | tuple): Size of the kernel:
+        stride (int | tuple): Stride of the filter. Default: 1:
+        dilation (int | tuple): Spacing between kernel elements.
+            Default: 1
```
Comment (Junjun2016) [inline mmcv/cnn/bricks/transformer.py:72]:
```suggestion
            Default: 1.
```

File: mmcv/cnn/bricks/transformer.py [hunk 12]
```diff
@@ -55,6 +59,332 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+
+    def __init__(self, kernel_size=1, stride=1, dilation=1, padding='corner'):
+
+        super(AdaptivePadding, self).__init__()
+
+        assert padding in ('same', 'corner')
+
+        kernel_size = to_2tuple(kernel_size)
+        stride = to_2tuple(stride)
+        padding = to_2tuple(padding)
```
Comment (Junjun2016) [inline mmcv/cnn/bricks/transformer.py:101]:
```suggestion
```

`padding` is `str`, do not need `to_2tuple`.

**Comment by ZwwWayne** [comment]
kindly ping @jshilong

**Comment by jshilong** [comment]
@Junjun2016 @ZwwWayne

**Comment by jshilong** [comment]
Ping @ZwwWayne

**Comment by Junjun2016** [comment]
Please fix the lint error.

File: mmcv/cnn/bricks/transformer.py [hunk 13]
```diff
@@ -55,6 +59,345 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+        if norm_cfg is not None:
+            self.norm = build_norm_layer(norm_cfg, embed_dims)[1]
+        else:
+            self.norm = None
+
+        if input_size:
+            input_size = to_2tuple(input_size)
+            # `init_out_size` would be used outside to
+            # calculate the num_patches
+            # when `use_abs_pos_embed` outside
```
Comment (Junjun2016) [inline mmcv/cnn/bricks/transformer.py:231]:
```suggestion
            # e.g. when `use_abs_pos_embed=True` outside
```

File: mmcv/cnn/bricks/transformer.py [hunk 14]
```diff
@@ -55,6 +59,345 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+        if self.norm is not None:
+            x = self.norm(x)
+        return x, out_size
+
+
+class PatchMerging(BaseModule):
+    """Merge patch feature map.
+
+    This layer groups feature map by kernel_size, and applies norm and linear
+    layers to the grouped feature map. Our implementation uses `nn.Unfold` to
```
Comment (Junjun2016) [inline mmcv/cnn/bricks/transformer.py:278]:
```suggestion
    layers to the grouped feature map (used in Swin). Our implementation uses `nn.Unfold` to
```

File: mmcv/cnn/bricks/transformer.py [hunk 15]
```diff
@@ -55,6 +59,345 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+                                                 f'input_size is ' \
+                                                 f'`Sequence` ' \
+                                                 f'but get {input_size}'
+
+        H, W = input_size
+        assert L == H * W, 'input feature has wrong size'
+
+        x = x.view(B, H, W, C).permute([0, 3, 1, 2])  # B, C, H, W
+        # Use nn.Unfold to merge patch. About 25% faster than original method,
+        # but need to modify pretrained model for compatibility
```
Comment (Junjun2016) [inline mmcv/cnn/bricks/transformer.py:378]: Maybe the comment should be moved to the right place (above code?).

File: mmcv/cnn/bricks/transformer.py [hunk 16]
```diff
@@ -55,6 +59,345 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+
+        if self.adap_padding:
+            x = self.adap_padding(x)
+            H, W = x.shape[-2:]
+
+        x = self.sampler(x)
+        # if kernel_size=2 and stride=2, x should has shape (B, 4*C, H/2*W/2)
```
Comment (Junjun2016) [inline mmcv/cnn/bricks/transformer.py:385]: Same here.

**Review by Junjun2016 — APPROVED** [review]
LGTM, except for some minor comments.

**Comment by jshilong** [comment]
Ping @ZwwWayne

File: mmcv/cnn/bricks/transformer.py [hunk 17]
```diff
@@ -275,7 +275,8 @@ class PatchMerging(BaseModule):
     """Merge patch feature map.
 
     This layer groups feature map by kernel_size, and applies norm and linear
-    layers to the grouped feature map. Our implementation uses `nn.Unfold` to
+    layers to the grouped feature map ((used in SwinTransformer)).
```
Comment (Junjun2016) [inline mmcv/cnn/bricks/transformer.py:278]:
```suggestion
    layers to the grouped feature map ((used in Swin Transformer)).
```

**Review by Junjun2016 — APPROVED** [review]

**Review by Junjun2016 — APPROVED** [review]

**Comment by jshilong** [comment]
Ping @ZwwWayne

File: mmcv/cnn/bricks/transformer.py [hunk 18]
```diff
@@ -55,6 +59,346 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+    We use a conv layer to implement PatchEmbed.
+
+    Args:
+        in_channels (int): The num of input channels. Default: 3
+        embed_dims (int): The dimensions of embedding. Default: 768
+        conv_type (str): The type of embedding-conv. Default: "Conv2d".
+        kernel_size (int): The kernel_size of embedding conv. Default: 16.
+        stride (int): The slide stride of embedding conv.
+            Default: 16.
+        padding (int | tuple | string ): The padding length of
```
Comment (ZwwWayne) [inline mmcv/cnn/bricks/transformer.py:163]: padding (int | tuple | string ) -> padding (int | tuple | string)

**Review by Junjun2016 — APPROVED** [review]
LGTM

File: mmcv/cnn/bricks/transformer.py [hunk 19]
```diff
@@ -55,6 +59,348 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+
+    def get_pad_shape(self, input_shape):
+        """Calculate the padding size of input.
+
+        Args:
+            input_shape (:obj:`torch.Size`): arrange as (H, W).
+
+        Returns:
+            Tuple[int]: The padding size along the
+                original H and W directions
```
Comment (zhouzaida) [inline mmcv/cnn/bricks/transformer.py:117]:
```suggestion
            Tuple[int]: The padding size along the
            original H and W directions
```

File: mmcv/cnn/bricks/transformer.py [hunk 20]
```diff
@@ -55,6 +59,348 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+    Args:
+        kernel_size (int | tuple): Size of the kernel. Default: 1.
+        stride (int | tuple): Stride of the filter. Default: 1.
+        dilation (int | tuple): Spacing between kernel elements.
+            Default: 1.
+        padding (str): Support "same" and "corner", "corner" mode
+            would pad zero to bottom right, and "same" mode would
+            pad zero around input. Default: "corner".
+    Example:
+        >>> kernel_size = 16
```
Comment (zhouzaida) [inline mmcv/cnn/bricks/transformer.py:80]:
```suggestion
            pad zero around input. Default: "corner".
            
    Example:
        >>> kernel_size = 16
```

File: mmcv/cnn/bricks/transformer.py [hunk 21]
```diff
@@ -55,6 +59,348 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+        """
+        Args:
+            x (Tensor): Has shape (B, C, H, W). In most case, C is 3.
+
+        Returns:
+            tuple: Contains merged results and its spatial shape.
+
+                - x (Tensor): Has shape (B, out_h * out_w, embed_dims)
+                - out_size (tuple[int]): Spatial shape of x, arrange as
+                    (out_h, out_w).
```
Comment (zhouzaida) [inline mmcv/cnn/bricks/transformer.py:261]:
```suggestion
            tuple: Contains merged results and their spatial shape.

            - x (Tensor): Has shape (B, out_h * out_w, embed_dims)
            - out_size (tuple[int]): Spatial shape of x, arrange as
              (out_h, out_w).
```

File: mmcv/cnn/bricks/transformer.py [hunk 22]
```diff
@@ -55,6 +59,348 @@ def build_transformer_layer_sequence(cfg, default_args=None):
...
+            x (Tensor): Has shape (B, H*W, C_in).
+            input_size (tuple[int]): The spatial shape of x, arrange as (H, W).
+                Default: None.
+
+        Returns:
+            tuple: Contains merged results and its spatial shape.
+
+                - x (Tensor): Has shape (B, Merged_H * Merged_W, C_out)
+                - out_size (tuple[int]): Spatial shape of x, arrange as
+                    (Merged_H, Merged_W).
```
Comment (zhouzaida) [inline mmcv/cnn/bricks/transformer.py:368]:
```suggestion
            tuple: Contains merged results and its spatial shape.

            - x (Tensor): Has shape (B, Merged_H * Merged_W, C_out)
            - out_size (tuple[int]): Spatial shape of x, arrange as
              (Merged_H, Merged_W).
```

**Comment by zhouzaida** [comment]
https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/bricks/__init__.py
https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/__init__.py
Should we add the module here?

**Comment by jshilong** [comment]
> https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/bricks/__init__.py https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/__init__.py Should we add the module here?

Done

File: mmcv/cnn/bricks/__init__.py [hunk 23]
```diff
@@ -31,5 +34,7 @@
     'UPSAMPLE_LAYERS', 'PLUGIN_LAYERS', 'Scale', 'ConvAWS2d', 'ConvWS2d',
     'conv_ws_2d', 'DepthwiseSeparableConvModule', 'Swish', 'Linear',
     'Conv2dAdaptivePadding', 'Conv2d', 'ConvTranspose2d', 'MaxPool2d',
-    'ConvTranspose3d', 'MaxPool3d', 'Conv3d', 'Dropout', 'DropPath'
+    'ConvTranspose3d', 'MaxPool3d', 'Conv3d', 'Dropout', 'DropPath',
+    'AdaptivePadding', 'PatchEmbed', 'PatchMerging', 'MultiheadAttention',
+    'FFN', 'BaseTransformerLayer', 'TransformerLayerSequence'
```
Comment (zhouzaida) [inline mmcv/cnn/bricks/__init__.py:39]: please add the Linear here.

**Comment by zhouzaida** [comment]
> > https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/bricks/__init__.py https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/__init__.py Should we add the module here?
> 
> Done

We should also update https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/__init__.py

**Comment by jshilong** [comment]
> https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/bricks/__init__.py https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/__init__.py Should we add the module here?

I do not recommend adding it because the modules in `transformer.py` are not basic enough and import some modules from
`mmcv.cnn`, adding it to `__init__.py` may cause the recursive import

**Review by ZwwWayne — APPROVED** [review]
